src/Language/Fixpoint/Solver/Interpreter.hs

```diff
@@ -0,0 +1,843 @@
+
+initCtx :: InstEnv a -> [(Expr,Expr)] -> ICtx
+initCtx _   es   = ICtx 
+  { icAssms  = mempty 
+  , icCands  = mempty 
+  , icEquals = S.fromList es
+  , icSolved = mempty
+  , icSimpl  = mempty 
+  , icSubcId = Nothing
+  }
+
+equalitiesPred :: S.HashSet (Expr, Expr) -> [Expr]
+equalitiesPred eqs = [ EEq e1 e2 | (e1, e2) <- S.toList eqs, e1 /= e2 ] 
+
+updCtxRes :: InstRes -> Maybe BindId -> ICtx -> (ICtx, InstRes) 
+updCtxRes res iMb ctx = (ctx, res')
+  where 
+    res' = updRes res iMb (pAnd $ equalitiesPred $ icEquals ctx)
+
+
+updRes :: InstRes -> Maybe BindId -> Expr -> InstRes
+updRes res (Just i) e = M.insertWith (error "tree-like invariant broken in ple. See https://github.com/ucsd-progsys/liquid-fixpoint/issues/496") i e res 
+updRes res  Nothing _ = res 
+
+
+---------------------------------------------------------------------------------------------- 
+-- | @updCtx env ctx delta cidMb@ adds the assumptions and candidates from @delta@ and @cidMb@ 
+--   to the context. 
+---------------------------------------------------------------------------------------------- 
+
+updCtx :: InstEnv a -> ICtx -> Diff -> Maybe SubcId -> ICtx 
+updCtx (InstEnv {..}) ctx delta cidMb 
+    = ctx { icCands  = S.fromList cands           <> icCands  ctx
+          , icEquals = initEqs                    <> icEquals ctx
+          , icSimpl  = M.fromList (S.toList sims) <> icSimpl ctx <> econsts
+          , icSubcId = cidMb -- fst <$> L.find (\(_, b) -> (head delta) `memberIBindEnv` (_cenv b)) ieCstrs
+          }                  -- ^^ eliminate if nothing broken ^^
+  where         
+    initEqs   = S.fromList $ concat [rewrite e rw | e  <- cands ++ (snd <$> S.toList (icEquals ctx))
+                                                  , rw <- snd <$> M.toList (knSims ieKnowl)]
+    cands     = concatMap (makeCandidates ieKnowl ctx) (rhs:es)
+    sims      = S.filter (isSimplification (knDCs ieKnowl)) (initEqs <> icEquals ctx)
+    econsts   = M.fromList $ findConstants ieKnowl es
+    (rhs:es)  = unElab <$> (eRhs : (expr <$> binds))
+    eRhs      = maybe PTrue crhs subMb
+    binds     = [ lookupBindEnv i ieBEnv | i <- delta ] 
+    subMb     = getCstr ieCstrs <$> cidMb
+
+
+findConstants :: Knowledge -> [Expr] -> [(Expr, Expr)]
+findConstants γ es = [(EVar x, c) | (x,c) <- go [] (concatMap splitPAnd es)]  
+  where 
+    go su ess = if ess == ess' 
+                  then su 
+                  else go (su ++ su') ess' 
+       where ess' = subst (mkSubst su') <$> ess
+             su'  = makeSu ess 
+    makeSu exprs  = [(x,c) | (EEq (EVar x) c) <- exprs 
+                           , isConstant (knDCs γ) c 
+                           , EVar x /= c ]
+
+makeCandidates :: Knowledge -> ICtx -> Expr -> [Expr]
+makeCandidates γ ctx expr 
+  = mytracepp ("\n" ++ show (length cands) ++ " New Candidates") cands
+  where 
+    cands = filter (\e -> isRedex γ e && not (e `S.member` icSolved ctx)) (notGuardedApps expr)
```

Comment:
This code has been fixed in the smt-based ple in https://github.com/ucsd-progsys/liquid-fixpoint/pull/504.